### DISCUSSION TITLE
Where can I find the code for the BlockCache B+ Tree?

### DISCUSSION THREAD

**Discussion by dannycho7** [facebook/CacheLib#240] [Q&A] [answered]
I am trying to learn more about CacheLib Navy and have been reading parts of the code. One part I am stuck on is finding where the B+ Tree code is. I am interested to see how this tiered mapping of the key hash to save bytes for metadata is implemented.

When I look at [`navy::Index`](https://github.com/facebook/CacheLib/blob/main/cachelib/navy/block_cache/Index.h#L190), the third and fourth byte of the 8-byte key hash appears to identify a bucket in a fixed size array and the last 4 bytes are used as the key for a hash map. Has the implementation changed since the paper and docs were written or maybe I'm misunderstanding the code?

For further context, I am referring to the B tree referenced in both the docs in cachelib.org and the OSDI '20 paper:

- From [docs](https://cachelib.org/docs/Cache_Library_Architecture_Guide/overview_a_random_walk/#blockcache-the-large-item-engine): "Each item is stored in a B tree index in memory. When a read is performed, the region that item is in is promoted in the FIFO queue in memory."
- From the [paper](https://www.usenix.org/system/files/osdi20-berg.pdf): "Keys are hashed to 8B. The first 4B identify the B+-
tree segment, and the second 4B are used as a key within in a tree segment to lookup a flash location".

**Comment by haowu14** [comment, marked as answer]
Yes we switched from B-tree to sparse_map after the paper was written. Benchmarking result at that time showed both latency and memory win by this switch.

**Reply by haowu14** [reply]
Just updated the doc. Thanks for bringing this up!

**Reply by dannycho7** [reply]
Thank you for the clarification and fix!

**Comment by pbhandar2** [comment]
Closing this discussion as it has been inactive. If this is still relevant, then please create a new discussion post and we will get back to you as soon as we can. Thank you!
